// hardhat-tracer/dist/src/wrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.wrapEthersProvider = exports.wrapHardhatProvider = void 0;
const ethers_1 = require("ethers");
const backwards_compatibility_1 = require("hardhat/internal/core/providers/backwards-compatibility");
const wrapper_1 = require("hardhat/internal/core/providers/wrapper");
const logs_1 = require("./logs");
/**
 * Wrapped provider which extends requests
 */
class TracerWrapper extends wrapper_1.ProviderWrapper {
    constructor(dependencies) {
        super(dependencies.provider);
        this.txPrinted = {};
        this.dependencies = dependencies;
    }
    async request(args) {
        // let result = (await (this.provider.request
        //   ? this._wrappedProvider.request(args)
        //   : // @ts-ignore
        //     this._wrappedProvider.send(args.method, args.params as any[]))) as any;
        var _a;
        let result = await this.dependencies.provider.send(args.method, args.params);
        if (result != null &&
            (args.method === "eth_sendTransaction" ||
                args.method === "eth_sendRawTransaction" ||
                args.method === "eth_getTransactionReceipt")) {
            let hash = result;
            let receipt;
            if (typeof result === "object" && result !== null) {
                hash = (_a = result.transactionHash) !== null && _a !== void 0 ? _a : result.hash;
                receipt = result;
            }
            else {
                receipt = await this.dependencies.provider.send("eth_getTransactionReceipt", [hash]);
            }
            if (!this.txPrinted[hash] && receipt !== null) {
                // console.log(hash);
                this.txPrinted[hash] = true;
                const dependenciesExtended = Object.assign(Object.assign({}, this.dependencies), { nameTags: Object.assign({}, this.dependencies.tracerEnv.nameTags) });
                // @ts-ignore
                await logs_1.printLogs(hash, receipt, dependenciesExtended);
            }
        }
        return result;
    }
}
/**
 * Add hardhat-tracer to your environment
 * @param hre: HardhatRuntimeEnvironment - required to get access to contract artifacts and tracer env
 */
function wrapHardhatProvider(hre) {
    const tracerProvider = new TracerWrapper({
        artifacts: hre.artifacts,
        tracerEnv: hre.tracer,
        provider: hre.network.provider,
    });
    const compatibleProvider = new backwards_compatibility_1.BackwardsCompatibilityProviderAdapter(tracerProvider);
    hre.network.provider = compatibleProvider;
    if (!hre.tracer) {
        hre.tracer = {
            nameTags: {},
            _internal: { printNameTagTip: undefined },
        };
    }
}
exports.wrapHardhatProvider = wrapHardhatProvider;
/**
 * Wrap hardhat-tracer over an ethers provider
 * @param provider an ethers provider to attach hardhat-tracer logic
 * @param artifacts hre.artifacts
 * @param tracerEnv hre.tracer
 * @returns
 */
function wrapEthersProvider(provider, artifacts, tracerEnv) {
    if (!tracerEnv) {
        tracerEnv = {
            nameTags: {},
            _internal: { printNameTagTip: undefined },
        };
    }
    const tracerProvider = new TracerWrapper({ provider, artifacts, tracerEnv });
    const compatibleProvider = new backwards_compatibility_1.BackwardsCompatibilityProviderAdapter(tracerProvider);
    return new ethers_1.ethers.providers.Web3Provider(compatibleProvider);
}
exports.wrapEthersProvider = wrapEthersProvider;
